### PR TITLE
script: add the option to run .bat, .cmd or .ps1 test scripts

### PR DESCRIPTION
**Pull request by gdams**

### CODE REVIEW

File: lib/lookup.js [hunk 1]
```diff
@@ -95,9 +95,12 @@ function resolve(context, next) {
         context.emit('verbose', 'lookup-install', rep.install);
         context.module.install = rep.install;
       }
-      if (rep.script) {
+      if (rep.script && process.platform != 'win32') {
```
Comment (bzoz) [inline lib/lookup.js]: nit: maybe `!==` ?

**Comment by GeorgeAdams95** [comment]
@TheAlphaNerd any objections?

File: lib/lookup.js (hunk 1 above)
Comment (gibfahn) [inline lib/lookup.js]: @TheAlphaNerd Is there a lint rule we could enable to check for `!=`?

File: lib/npm/test.js [hunk 2]
```diff
@@ -28,6 +28,9 @@ function authorName(author) {
 }
 
 function test(context, next) {
+  if (context.options.script) {
+    context.module.script = context.options.script;
```
Comment (gibfahn) [inline lib/npm/test.js]: This is for the `citgm` (as opposed to `citgm-all`) case right? Could you add a comment in here noting that?

File: lib/npm/script/run.js [hunk 3]
```diff
@@ -22,7 +22,16 @@ function runScript(context, script, msg, next) {
       return next(err);
     }
     var options = createOptions(context.path, context);
-    var proc = spawn(_path, [], options);
+    var scriptExtension = path.extname(_path);
+    var proc = '';
+    if ((scriptExtension === '.bat' || scriptExtension === '.cmd') && process.platform === 'win32' || scriptExtension === '.sh' && process.platform != 'win32') {
```
Comment (gibfahn) [inline lib/npm/script/run.js]: Would be good to break this over multiple lines

Comment (gibfahn) [inline lib/npm/script/run.js]: Also `!=` -> `!==`

Comment (gibfahn) [inline lib/npm/script/run.js]: Maybe `['.bat','.cmd'].indexOf(scriptExtension) !== -1` rather than `(scriptExtension === '.bat' || scriptExtension === '.cmd')`? Not that much better tbh, so up to you.

**Comment by GeorgeAdams95** [comment]
@gibfahn updated PTAL

File: lib/create-options.js [hunk 4]
```diff
@@ -10,6 +10,7 @@ function createOptions(cwd, context) {
   options.env = Object.create(process.env);
   options.env['npm_loglevel'] = 'error';
   options.env['npm_config_tmp'] = context.npmConfigTmp;
+  options.shell = true;
```
Comment (TheAlphaNerd) [inline lib/create-options.js]: do we need to this for just windows? Is there a side effect to have this always set?

File: README.md [hunk 5]
```diff
@@ -115,6 +115,7 @@ For syntax, see [lookup.json](./lib/lookup.json), the available attributes are:
 "repo": "https://github.com/pugjs/jade" - Use a different github repo
 "skipAnsi": true             Strip ansi data from output stream of npm
 "script": /path/to/script | https://url/to/script - Use a custom test script
+"scriptWin": \path\to\script | https://url/to/script - Use a custom windows test script
```
Comment (TheAlphaNerd) [inline README.md]: Are we going to often need a different script just for windows?

 I'm not a very big fan of this, but understand if it is necessary

File: lib/npm/script/run.js [hunk 6]
```diff
@@ -22,7 +22,17 @@ function runScript(context, script, msg, next) {
...
+    var proc = '';
+    if ((['.bat','.cmd'].indexOf(scriptExtension) !== -1 && process.platform === 'win32') ||
+     (scriptExtension === '.sh' && process.platform !== 'win32')) {
+      proc = spawn(_path, [], options);
+    } else if (scriptExtension === '.ps1') {
+      proc = spawn('powershell.exe',[_path], options);
+    } else {
+      context.emit('fail', 'verbose', 'unsupported', scriptExtension + ' not supported');
+      return next(Error(msg));
+    }
```
Comment (TheAlphaNerd) [inline lib/npm/script/run.js]: is it right to assume that this now enforces an extension on all scripts? If so this will be a semver major change.

File: lib/npm/test.js [hunk 7]
```diff
@@ -28,6 +28,10 @@ function authorName(author) {
 }
 
 function test(context, next) {
+  if (context.options.script) {
+    // This handles the citgm.js case
+    context.module.script = context.options.script;
```
Comment (TheAlphaNerd) [inline lib/npm/test.js]: Is this related to the script changes that just happened? Why is this line necessary

File: README.md (hunk 5 above)
Comment (GeorgeAdams95) [inline README.md]: We will need it for our test suite to pass the script tests

File: lib/npm/test.js (hunk 7 above)
Comment (GeorgeAdams95) [inline lib/npm/test.js]: okay so citgm.js puts script in `context.options.script` and citgm-all.js put script in `context-module-script` This was broken in a recent PR but I couldn't find which one

Comment (TheAlphaNerd) [inline lib/npm/test.js]: can you please send a different PR to fix this. Potentially fixing in CITGM so the behavior is the same and this catch is not neccessary

File: README.md (hunk 5 above)
Comment (gibfahn) [inline README.md]: I guess if `script` can be a string or an object that's semver-minor? I agree allowing it to be an object gives a lot more freedom.

Could we maybe use the same options we use for skipping tests? That way you could pass different scripts for different versions of node or whatever (and the syntax would match, consistency is always nice).

File: lib/npm/test.js (hunk 7 above)
Comment (bzoz) [inline lib/npm/test.js]: Is https://github.com/nodejs/citgm/pull/245 the offending PR?

File: README.md (hunk 5 above)
Comment (GeorgeAdams95) [inline README.md]: the tests that we have that run shell scripts and check the return code will fail on windows unless a batch equivalent is added

File: lib/npm/script/run.js (hunk 6 above)
Comment (gdams) [inline lib/npm/script/run.js]: @MylesBorins the extension is no longer required, I have updated it to behave the same as before when there is no file path

**Comment by gdams** [comment]
@MylesBorins I have moved this to `semver minor` as there is no breaking change anymore

**Comment by bzoz** [comment]
Maybe we could drop support for everything besides node.js scripts? It would be simple and portable.

**Comment by MylesBorins** [comment]
> Maybe we could drop support for everything besides node.js scripts

We currently are not using anything but npm scripts ATM. @gibfahn @gdams do we have any custom scripts in the IBM infra

**Comment by gibfahn** [comment]
AFAIK we aren't using custom scripts in the IBM infra at all.

**Comment by gibfahn** [comment]
https://github.com/nodejs/citgm/issues/246 is an example of a use for custom scripts

**Comment by gdams** [comment]
Yes to confirm, we don't use any custom scripts in the IBM lookup

**Comment by gdams** [comment]
closing this as we are dropping script support in v2.0.0
